### ISSUE TITLE
[FEATURE] improved list/set management

### ISSUE THREAD

**Issue by vicariousdrama** [vitorpamplona/amethyst#815]
**Is your feature request related to a problem? Please describe.**

Lists are a key method by which users organize content. The current support for lists in amethyst requires users to leave and use another client that offers the ability to work with lists beyond uncategorised bookmarks and mute lists

**Describe the solution you'd like**

As a user, I want to have full control of my lists within the mobile application.

- All of my lists that are not marked as deleted should appear in the drop-down to use for filtering.
- A user should be able to **create** a new named list/set (name/title)
- A user should be able to **rename** a list/set (to correct for spelling errors)
- A user should be able to **copy** a list as a new list/set
- A user should be able to **add a pubkey** (p tag) to a list from the user's profile, or from an event they published (e.g., menu action to `Add author to list`, then present a submenu or modal allowing selection of an existing list or creating a new list)
- A user should be able to **add an event** (e tag) id to a list from the event when viewed (e.g. menu action to `Add event to list`, then present a submenu or modal allowing selection of an existing list or creating a new list)
- A user should be able to **see all items** on their list, and an indicator of the kind
- A user should be able to **add a URL** (r tag) to a list from where a list is being managed.
- A user should be able to **delete** an item from the list where its being managed.
- A user should be able to **reorder** items in their list (can be implemented as move up/move down or drag and drop or other suitable approach)

Nice to haves, but not required
- ability to set and update a picture (image) for a list, and its about (description) tag
- listing lists of followers and being able to make a fork of them

In a nutshell, a user should have full CRUD over their personal lists within Amethyst

**Bounty (in Bitcoin sats) offered for the implementation**

I (npub1yx6pjypd4r7qh2gysjhvjd9l2km6hnm4amdnjyjw3467fy05rf0qfp7kza) am offering 1_000_000 sats for this feature to be added for kinds 10003, 30000, 30003 within constraints of allowed tags

Please advise if this should be broken down into multiple issues and bounties.

Inconsistencies and shortcomings of NIP51 may need to be addressed prior to implementation so that clients can get on the same page

**Comment by vicariousdrama** [comment]
I (npub1yx6pjypd4r7qh2gysjhvjd9l2km6hnm4amdnjyjw3467fy05rf0qfp7kza) am increasing my bounty offer to 2_000_000 sats for this feature to be added

**Comment by KotlinGeekDev** [comment]
Well, I'm interested in this. Please @vicariousdrama @vitorpamplona do you think it should be similar to the relay editing UX?

**Comment by vitorpamplona** [comment]
Yep, but also the "Follow" button on the user profile should open a list of lists to add the user in, multi-selection. 

Lots of new screens to manage lists. Feel free to start small and keep adding to the solution on different PRs.

**Comment by vicariousdrama** [comment]
Yes. It may be more feasible to break it down into phases. I just know that myself and plenty of others really really want better lists directly within the app, so the key is moving in that direction.

The follow button on user profile, and on the context menu on a post would ideally allow for putting the user in an existing list or starting a new list with that person

**Comment by KotlinGeekDev** [comment]
Hello @vicariousdrama . Sorry you had to wait long for this feature. I will get to it very soon(this week?). There are 2 other issues(unrelated to this one) that I'll finish before tackling this one. Thanks for your patience. :pray:

**Comment by KotlinGeekDev** [comment]
Please @vicariousdrama @vitorpamplona what do you think of 'add event to list' and 'add URL to list', being replaced by "Bookmark event/post" and "Bookmark URL/link" respectively?

**Comment by vicariousdrama** [comment]
I think the terminology changes for labeling is fine, so long as its clear in the flow that a bookmark is going to go to a user selectable list (30003 kinds are the user curated Bookmark Lists), or default list (10003 is the default Bookmark List)

**Comment by KotlinGeekDev** [comment]
> I think the terminology changes for labeling is fine, so long as its clear in the flow that a bookmark is going to go to a user selectable list (30003 kinds are the user curated Bookmark Lists), or default list (10003 is the default Bookmark List)

Got it. So a kind 30003 set could be called "Labeled Bookmark List", and the kind 10003 list could be unchanged; "Default Bookmark List".

**Comment by KotlinGeekDev** [comment]
While working on this, I noticed that the 'd' tag is being used as a form of label, though NIP-51 also provides options for putting a title. Some duplicate the value across the d tag, and title. Should we go for duplication, or just use the 'd' tag? cc @vitorpamplona

**Comment by KotlinGeekDev** [comment]
Also, @vicariousdrama @vitorpamplona  would it better to default to a private list(especially for follows ) when creating one, or always show all the options?

**Comment by vitorpamplona** [comment]
No duplication is needed. Clients should render the `d` tag if `title` is not available. 

Good question about putting in the private or public part of the list. Lots of people want private follows, so we might need to figure out a UI to do both.

**Comment by vitorpamplona** [comment]
You could use a UUID for d and put a name on title so that if users want to change the name, they can change title without changing the d tag, which would generate a new list on Nostr

**Comment by KotlinGeekDev** [comment]
> You could use a UUID for d and put a name on title so that if users want to change the name, they can change title without changing the d tag, which would generate a new list on Nostr

Exactly. That was the initial idea I had until I saw the duplications.

**Comment by KotlinGeekDev** [comment]
@vitorpamplona in PeopleListEvent, we use `name` and assign it to the `d` tag. :sweat_smile:

**Comment by vitorpamplona** [comment]
Oh yeah, Nostr moved from `title` to `name` some time ago.

**Comment by KotlinGeekDev** [comment]
> Oh yeah, Nostr moved from `title` to `name` some time ago.

Sorry, I meant that it's what I saw in the code. :sweat_smile:

**Comment by KotlinGeekDev** [comment]
Please @vitorpamplona there seems to be no way to pull addressable events from the cache without specifying a `d` tag value. :thinking:

**Comment by vitorpamplona** [comment]
Yeah, we never really had that need so the current way it's done is more of a hack. Check the "LoadStatuses", which loads a list of StatusEvents. It's not great (refactoring is welcome), but you can copy it.

**Comment by KotlinGeekDev** [comment]
> Yeah, we never really had that need so the current way it's done is more of a hack. Check the "LoadStatuses", which loads a list of StatusEvents. It's not great (refactoring is welcome), but you can copy it.

Got it. I found another 'hacky' way: filtering the addressables from the cache that match the kind we want.

**Comment by KotlinGeekDev** [comment]
> Got it. I found another 'hacky' way: filtering the addressables from the cache that match the kind we want.

@vitorpamplona  I think I might put the solution above in `getOrCreateAddressableNote(ATag)` , so as to maintain some compatibility with the present code.

**Comment by KotlinGeekDev** [comment]
> Yep, but also the "Follow" button on the user profile should open a list of lists to add the user in, multi-selection.
> 
As for this @vitorpamplona , could we use a UX similar to Nostrudel, whereby we we have a dropdown next to the Follow button?

**Comment by vitorpamplona** [comment]
Sure!

**Comment by KotlinGeekDev** [comment]
> Sure!

Got it.

**Comment by KotlinGeekDev** [comment]
Please @vicariousdrama does the bounty still hold? 😅

**Comment by vicariousdrama** [comment]
Yes. The bounty still holds. I'm offering 2 million sats for completion of all the base features, and can be broken down into chunks for progression through releases.  I do see an open pull request with over 50 commits in #1169.

**Comment by KotlinGeekDev** [comment]
> Yes. The bounty still holds. 

Ok sir, thanks.

**Comment by KotlinGeekDev** [comment]
>  I do see an open pull request with over 50 commits in [#1169](https://github.com/vitorpamplona/amethyst/pull/1169).

Still working on it. 🙏

**Comment by vitorpamplona** [comment]
Yes, we can change it to dTag to be very explicit.

**Comment by KotlinGeekDev** [comment]
> Yes, we can change it to dTag to be very explicit.

Got it.

**Comment by KotlinGeekDev** [comment]
Follow sets done. Please @vicariousdrama take a look at the demo I linked in #1169 .

**Comment by vicariousdrama** [comment]
Thanks @KotlinGeekDev . I've just reviewed the demo video and it looks functional.

**Comment by KotlinGeekDev** [comment]
As for general bookmarks, I realised that Amethyst need to migrate to the new kind recommended by NIP-51. @vicariousdrama the UI for general bookmarks(kind 10,003) is already what Amethyst uses for Bookmarks. :) It will just need migrating users' lists from old kind to new kind.

**Comment by KotlinGeekDev** [comment]
Also, I came across an interesting limitation of Nostr lists/sets: when a private list/set becomes empty, there is no difference between it and an empty public list/set(since the `content` field of the private set is empty). 
What I did with Amethyst is make it appear public(in the UI) when it becomes empty. But I wonder if there's a better solution... @vitorpamplona

**Comment by vitorpamplona** [comment]
hum... I am not sure about separating lists in either public or private makes sense. I think all lists have a public component and a private component. People can always mix it. I think users need to see which entries are public and which ones are private, but they don't need to know if the whole list is private or public.

**Comment by KotlinGeekDev** [comment]
> hum... I am not sure about separating lists in either public or private makes sense. I think all lists have a public component and a private component. People can always mix it. I think users need to see which entries are public and which ones are private, but they don't need to know if the whole list is private or public.

Got it.

**Comment by vicariousdrama** [comment]
> As for general bookmarks, I realised that Amethyst need to migrate to the new kind recommended by NIP-51. [@vicariousdrama](https://github.com/vicariousdrama) the UI for general bookmarks(kind 10,003) is already what Amethyst uses for Bookmarks. :) It will just need migrating users' lists from old kind to new kind.

There may need to be a architectural discussion around best practice for kind 10003 -> 30003.  

I personally think of 10003 as the default catchall, and a natural root/top for any potential hierarchy as a pubkey can only have 1.  Kind 30003 could then be referenced via a tags in the kind 10003 to build out a folder hierarchy, or not referenced at all. Under this model, no direct migration from 10003 -> 30003 would be necessary, but support to `move` or `copy` an entry from one list to another would be useful.  Those actions can be considered future enhancements and aren't necessary for this particular issue or bounty payout. 

I don't think data should be migrated from 10003 to 30003 as it would break other apps that use global 10003 but are not yet supportive of labeled 30003 lists.  The logic for managing entries in bookmark related lists (10003, 30003), and user related lists (30000) should be fairly consistent.

**Comment by KotlinGeekDev** [comment]
> > As for general bookmarks, I realised that Amethyst need to migrate to the new kind recommended by NIP-51. [@vicariousdrama](https://github.com/vicariousdrama) the UI for general bookmarks(kind 10,003) is already what Amethyst uses for Bookmarks. :) It will just need migrating users' lists from old kind to new kind.
> 
> There may need to be a architectural discussion around best practice for kind 10003 -> 30003.
> 
> I personally think of 10003 as the default catchall, and a natural root/top for any potential hierarchy as a pubkey can only have 1. Kind 30003 could then be referenced via a tags in the kind 10003 to build out a folder hierarchy, or not referenced at all. Under this model, no direct migration from 10003 -> 30003 would be necessary, but support to `move` or `copy` an entry from one list to another would be useful. Those actions can be considered future enhancements and aren't necessary for this particular issue or bounty payout.
> 
> I don't think data should be migrated from 10003 to 30003 as it would break other apps that use global 10003 but are not yet supportive of labeled 30003 lists. The logic for managing entries in bookmark related lists (10003, 30003), and user related lists (30000) should be fairly consistent.

Yeah @vicariousdrama you're right about that. Sorry, I forgot to mention that Amethyst still uses the old kind for bookmarks(kind 30,001 with the "bookmark" tag). So, we'll migrate from that old way to the recommended kind 10,003 for bookmarks.
As for kind 30,003 bookmark sets, I think I will put them in the same section as follow sets, though I am not sure yet. Putting them in the bookmarks section seems better, but requires a quite big refactor there. But we'll see how to adjust that.

**Comment by KotlinGeekDev** [comment]
Bookmark lists(kind 30003), or bookmark groups done. Please @vicariousdrama see #1586 .

**Comment by KotlinGeekDev** [comment]
I think this issue can now be closed @vitorpamplona .
